alf/environments/wrappers.py

```diff
@@ -102,3 +104,200 @@ def step(self, action):
+    """
+    def __init__(self,
+                 env,
+                 frame_skip=4,
+                 noop_max=30,
+                 terminal_on_life_loss=False,
+                 screen_size=84):
+        """Constructor for an Atari 2600 preprocessor.
+
+        Args:
+        environment: Gym environment whose observations are preprocessed.
+        frame_skip: int, the frequency at which the agent experiences the game.
+        terminal_on_life_loss: bool, If True, the step() method returns
+            is_terminal=True whenever a life is lost. See Mnih et al. 2015.
+        screen_size: int, size of a resized Atari 2600 frame.
+
+        Raises:
+        ValueError: if frame_skip or screen_size are not strictly positive.
+        """
+        super().__init__(env)
+        if frame_skip <= 0:
+            raise ValueError(
+                'Frame skip should be strictly positive, got {}'.format(frame_skip))
+        if screen_size <= 0:
+            raise ValueError('Target screen size should be strictly positive, got {}'
+                            .format(screen_size))
+
+        self.terminal_on_life_loss = terminal_on_life_loss
+        self.frame_skip = frame_skip
+        self.screen_size = screen_size
+        self.noop_max = noop_max
+        assert env.unwrapped.get_action_meanings()[0] == "NOOP"
+
+        obs_dims = self.env.observation_space
+        # Stores temporary observations used for pooling over two successive
+        # frames.
+        self.screen_buffer = [
+            np.empty((obs_dims.shape[0], obs_dims.shape[1]), dtype=np.uint8),
+            np.empty((obs_dims.shape[0], obs_dims.shape[1]), dtype=np.uint8)
+        ]
+
+        self.observation_space = gym.spaces.Box(
+            low=0,
+            high=255,
+            shape=(self.screen_size, self.screen_size, 1),
+            dtype=np.uint8)
+
+        self.game_over = True
+        self.lives = 0  # Will need to be set by reset().
+
+    def reset_with_random_noops(self):
+        self.env.reset()
+        n_noops = self.env.unwrapped.np_random.randint(
+            1, self.noop_max + 1)
+        for _ in range(n_noops):
+            _, _, game_over, _ = self.env.step(0)
+            if game_over:
+                self.env.reset()
+
+    def fire_on_reset(self):
+        # The following code is from https://github.com/openai/gym/...
+        # ...blob/master/gym/wrappers/atari_preprocessing.py
+        action_meanings = self.env.unwrapped.get_action_meanings()
+        if action_meanings[1] == 'FIRE' and len(action_meanings) >= 3:
+            self.env.step(1)
+            self.env.step(2)
+
+    def reset(self):
+        """Resets the environment.
+
+        Returns:
+        observation: numpy array, the initial observation emitted by the
+            environment.
+        """
+        if self.game_over:
+            self.reset_with_random_noops()
+            self.lives = self.env.ale.lives()
+            self.game_over = False
+        else:
```

Comment:
is `self.game_over`  always `True` when reset is called ?

Reply:
If terminate_on_life_loss is True, when a life is lost, step() returns `terminal`=True to indicate that reset() should be called. self.game_over is maintained separately to decide if the game has really ended.